SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc

```diff
@@ -0,0 +1,453 @@
+  edm::LogInfo("CaloTruthAccumulator") << "Adding " << output_.pSimClusters->size() 
+				       << " SimParticles and " << output_.pCaloParticles->size()
+				       << " CaloParticles to the event.";
+
+  // now we need to normalize the hits and energies into hits and fractions
+  // (since we have looped over all pileup events)
+
+  for( auto& sc : *(output_.pSimClusters) ) {
+    auto hitsAndEnergies = sc.hits_and_fractions();
+    sc.clearHitsAndFractions();
+    for( auto& hAndE : hitsAndEnergies ) {
+      const float fraction = hAndE.second/m_detIdToTotalSimEnergy[hAndE.first];
+      sc.addRecHitAndFraction(hAndE.first,fraction);
+    }
+  }
+  
+  // save the SimCluster orphan handle so we can fill the calo particles
+  auto scHandle = event.put( std::move(output_.pSimClusters), "MergedCaloTruth" );
+  
+  // now fill the calo particles
+  for( unsigned i = 0; i < output_.pCaloParticles->size(); ++i ) {
+    auto& cp = (*output_.pCaloParticles)[i];
+    for( unsigned j = m_caloParticles.sc_start_[i]; j < m_caloParticles.sc_stop_[i]; ++j ) {
+      edm::Ref<SimClusterCollection> ref(scHandle,j);
+      cp.addSimCluster(ref);
+    }
+  }  
+
+  event.put( std::move(output_.pCaloParticles), "MergedCaloTruth" );
+
+  calo_particles().swap(m_caloParticles);
+
+  std::unordered_map<Index_t,float>().swap(m_detIdToTotalSimEnergy);
+
+  std::unordered_map<Barcode_t,Index_t>().swap(m_genParticleBarcodeToIndex);
+  std::unordered_map<Barcode_t,Index_t>().swap(m_simTrackBarcodeToIndex);
+  std::unordered_map<Barcode_t,Index_t>().swap(m_genBarcodeToSimTrackIndex);
+  std::unordered_map<Barcode_t,Index_t>().swap(m_simVertexBarcodeToIndex);
+
+  std::unordered_multimap<Index_t,Index_t>().swap(m_detIdToCluster);
+  std::unordered_multimap<Barcode_t,Index_t>().swap(m_simHitBarcodeToIndex);
+  std::unordered_multimap<Barcode_t,Barcode_t>().swap(m_simVertexBarcodeToSimTrackBarcode);
+  std::unordered_map<Barcode_t,Barcode_t>().swap(m_simTrackBarcodeToSimVertexParentBarcode);
+  std::unordered_multimap<Barcode_t,Index_t>().swap(m_simTrackToSimVertex);
+  std::unordered_multimap<Barcode_t,Index_t>().swap(m_simVertexToSimTrackParent); 
+  std::vector<Barcode_t>().swap(m_simVertexBarcodes);
+  std::unordered_map<Index_t,float>().swap(m_detIdToTotalSimEnergy);
+}
+
+template<class T> 
+void CaloTruthAccumulator::accumulateEvent( const T& event, 
+					    const edm::EventSetup& setup, 
+					    const edm::Handle< edm::HepMCProduct >& hepMCproduct) {
+  //
+  // Get the collections
+  //
+  
+  edm::Handle< std::vector<reco::GenParticle> > hGenParticles;
+  edm::Handle< std::vector<int> > hGenParticleIndices;
+  
+  event.getByLabel( simTrackLabel_, hSimTracks );
+  event.getByLabel( simVertexLabel_, hSimVertices );
+  
+  try {
+    event.getByLabel( genParticleLabel_, hGenParticles );
+    event.getByLabel( genParticleLabel_, hGenParticleIndices );
+  } catch( cms::Exception& exception ) {
```

Comment:
ah, no, there we go, exception on the 5th event. I'll see what's going on.

-L

On Sun, Aug 7, 2016 at 11:09 AM, Lindsey Gray <lindsey.gray@gmail.com>
wrote:

> weird, I am running right now with no crash (over xrootd on a number of
> files)
>
> On Sun, Aug 7, 2016 at 11:03 AM, Slava Krutelyov <notifications@github.com
> > wrote:
>
>> In SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc
>> <https://github.com/cms-sw/cmssw/pull/15293#discussion_r73804501>:
>>
>> > +					    const edm::EventSetup& setup,
>> > +					    const edm::Handle< edm::HepMCProduct >& hepMCproduct) {
>> > +  //
>> > +  // Get the collections
>> > +  //
>> > +
>> > +  edm::Handle< std::vector<reco::GenParticle> > hGenParticles;
>> > +  edm::Handle< std::vector<int> > hGenParticleIndices;
>> > +
>> > +  event.getByLabel( simTrackLabel_, hSimTracks );
>> > +  event.getByLabel( simVertexLabel_, hSimVertices );
>> > +
>> > +  try {
>> > +    event.getByLabel( genParticleLabel_, hGenParticles );
>> > +    event.getByLabel( genParticleLabel_, hGenParticleIndices );
>> > +  } catch( cms::Exception& exception ) {
>>
>> actually, looks like over-simplified the command: the one with 1 minbias
>> input file works OK.
>> After I switch to two file:8242B8BA-0053-E611-AEEF-0
>> 025905B85C0.root,file:101FD8D3-F852-E611-865C-0025905A605E.root
>> the crash is on the first event
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cms-sw/cmssw/pull/15293/files/838ad774de1cd5bf024063225a320c946e41a95b#r73804501>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABBMORKLlc0Gi0JEBnhudaqZ0pmz7i4zks5qdgHPgaJpZM4JVi3o>
>> .
>>
>
>